teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,924 @@
+Additionally, optional by default behavior will be useful if we pursue
+[the future option of supporting pre-defined schemas](#potential-future-work): optional by default will allow Tasks to
+refer to schemas for their params, and additive changes can be made to those schemas without requiring all Tasks using
+them to be updated.
+
+### Notes/Caveats
+
+[See also TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477) for notes and caveats
+specific to supporting json results.
+
+* Since [json only supports string types for keys (aka "names")](https://datatracker.ietf.org/doc/html/rfc7159#section-4)
+  we would only support string types for keys
+* What if a Pipeline tries to use a object key that doesn't exist in a param or result?
+    * Some of this could be caught at creation time, i.e. pipeline params would declare the keys each object param will
+      contain, and we could validate at creation time that those keys exist
+    * For invalid uses that can only be caught at runtime (e.g. after fetching the task spec), the PipelineRun would
+      fail
+
+### Risks and Mitigations
+
+[See TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477) for risks and mitigations
+specific to expanding ArrayOrString to support more types, including size limitations caused by our implementation of
+results ([pipelines#4012](https://github.com/tektoncd/pipeline/issues/4012)).
+
+### User Experience
+
+* Using the JSON Schema syntax to describe the structure of the object is a bit verbose but looking at our
+  [Alternatives](#alternatives) this seems like the best option assuming we want to support more complex types
+  in the future.
+
+## Design Details
+
+* [Declaring object results and params](#declaring-object-results-and-params)
+* [Emitting object results](#emitting-object-results)
+* [Using objects in variable replacement](#using-objects-in-variable-replacement)
+
+### Declaring object results and params
+
+We would use [JSON object schema syntax](https://json-schema.org/understanding-json-schema/reference/object.html) to
+declare the structure of the object.
+
+Declaring defaults for parameters that are of type object would follow the pattern we have already established with
+[array param defaults](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#specifying-parameters) which is
+to declare the value in the expected format in yaml (which is treated as json by the k8s APIs).
+
+Params example with default value:
+
+```yaml
+  params:
+    - name: pull_remote
+      description: JSON Schema has no "description" fields, so we'd have to include documentation about the structure in this field
+      type: object
+      properties:
+        url: {
+          type: string
+        }
+        path: {
+          type: string
+        }
+      default:
+        url: https://github.com/somerepo
+        path: ./my/directory/
```

Comment:
Are default applied individually or as a whole?
I.e. is the param is specified with `url` but `path` is not set, will `path` be taken from the default? 

My expectation would be no, so defaults are only applied to the whole object when it's not specified at all.
That might imply that we have to validate default values too. For instance if `path` was a required field, and the default did not include it, we should fail validation.

Reply:
good point, i agree! i'll add this in the implementable PR